enhancements/user-workload-monitoring.md

```diff
@@ -0,0 +1,275 @@
+- [PodMonitor](https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#podmonitor)
+- [PrometheusRule](https://github.com/openshift/prometheus-operator/blob/master/Documentation/api.md#prometheusrule)
+
+Allow the correlation between app and cluster centric metrics.
+
+### Non-Goals
+
+Out of scope are:
+
+- Definition of Dashboards.
+- Deployment of Prometheus instances in user namespaces.
+- Deployment of Alertmanager instances in user namespaces.
+
+## Proposal
+
+![](user-monitoring-deployment.png)
+
+The existing cluster monitoring stack is going to be extended with additional Prometheus (P) servers
+and one additional Prometheus Operator (PO) reconciling user namespaces
+as well as the newly added Prometheus servers.
+
+The cluster Prometheus Operator (PO) continues to operate the cluster-level Prometheus.
+We will continue to use a centralized Alertmanager (A) cluster.
+The existing Alertmanager cluster will aggregate both user workload alerts as well as cluster alerts. Tenancy will be achieved by forcing specific labels onto metrics and alerts by which the central Alertmanager config can route similarly to how tenancy is achieved at the query level (see below).
+
+Querying data from a single, multi-tenant interface is done using the Thanos Querier (TQ) component.
+Tenancy is enforced at the prometheus query layer. This is achived by leveraging the existing topology using kube-rbac-proxy [1] and prom-label-proxy [2].
+
+Differences between cluster- and user-level is mainly upgrade guarantees. The cluster-level Prometheus stack is an integral part of OpenShift and its reliability is important, keeping the stack immutable and fully testable by us allows us to rely on it to a higher degree. As we have no such possibility for user defined configurations and scrape targets, we can't provide the same guarantees.
+
+There are edge cases where an automated upgrade is not possible in the face of user workload monitoring.
+If users declare i.e. queries in ServiceMonitors which break in newer Prometheus versions
+we most likely cannot perform an automated migration. Such migration without knowledge of user intention behind each query would be exceptionally difficult.
+
+Instead, the "upgradable" condition in the ClusterOperator will be set to false.
+Routing configurations for alerts to be sent to specific teams need to be communicated/configured by an infrastructure owner using the existing Alertmanager stack that is part of cluster-monitoring.
+We will look into extension mechanisms post GA.
+With any OpenShift deployment, we only install the cluster-level part
+and allow infrastructure owner to "enable" the addition of user workload monitoring.
+
+### User Stories
+
+
+#### US1
+
+As a service owner, I’d like to configure a ServiceMonitor CRD so that Red Hat’s supported monitoring solution is able to scrape metrics for my service.
```

Comment:
I urge you to think through the RBAC impacts of allowing a customer access to these CRs and if there could be any unintended side-effects of opening up this access (would it give them access to break the cluster-level prom?).

Reply:
since this stack is independent from the cluster-level prom, and we will restrict custom resources for user workload monitoring, i am less concerned here.

Reply:
Restrict them to what exactly?  Will you ship clusterroles/bindings?
I guess I need more detail on how that will be implemented.

Reply:
The operator will literally only watch the `openshift-user-monitoring` for the Prometheus objects, so even if a user manages to bypass or escalate RBAC it still won't do anything.

Reply:
Users won't need clusterroles, but just regular roles to create `ServiceMonitor`, `PodMonitor`, and `PrometheusRule` custom resources in their namespaces. If users create `Prometheus` or `Alertmanager` custom resources in their namespace (given they have RBAC to do so), they will be ignored by the user workload monitoring prometheus operator anyways.

Due to the described features (allow-list, deny-list), these custom resources will only be reconciled by the user workload prometheus operator, not the existing cluster monitoring prometheus operator.